relay-server/src/services/projects/cache/state.rs

```diff
@@ -111,9 +95,43 @@ impl ProjectStore {
             histogram(RelayHistograms::ProjectStateCacheSize) = self.private.len() as u64,
             storage = "private"
         );
-        metric!(counter(RelayCounters::EvictingStaleProjectCaches) += evicted as u64);
 
-        evicted
+        new_fetch
+    }
+
+    /// Waits for the next scheduled eviction and returns an [`Eviction`] token.
+    ///
+    /// The returned [`Eviction`] token must be immediately turned in using [`Self::evict`].
+    ///
+    /// The returned future is cancellation safe.
+    pub async fn next_eviction(&mut self) -> Option<Eviction> {
+        if self.evictions.is_empty() {
+            return None;
+        }
+        self.evictions.next().await.map(Eviction)
+    }
+
+    /// Evicts a project using an [`Eviction`] token returned from [`Self::next_eviction`].
+    pub fn evict(&mut self, Eviction(project_key): Eviction) {
+        // Remove the private part.
+        let Some(private) = self.private.remove(&project_key) else {
+            // Not possible if all invariants are upheld.
+            debug_assert!(false, "no private state for evicition");
```

Comment:
Should this be an `.expect()` call instead? Panicking in production is bad, but if it really cannot happen I think it's justified.

If there's a chance this happens in production, but is unlikely to happen during tests, I would log an error instead.

Reply:
I don't want to panic, because currently this completely breaks Relay (because it aborts the service task). But the error case here isn't so bad, this basically just means the project we tried to evict didn't exist anymore (we wanted to evict it anyways).